File: AMF5/src/test/java/SchedulerTest.java
import amf.apicontract.client.platform.AMFConfiguration;
import amf.apicontract.client.platform.RAMLConfiguration;
import amf.apicontract.client.platform.model.domain.EndPoint;
import amf.apicontract.client.platform.model.domain.Operation;
import amf.apicontract.client.platform.model.domain.Payload;
import amf.apicontract.client.platform.model.domain.Request;
import amf.apicontract.client.platform.model.domain.api.WebApi;
import amf.core.client.common.validation.ValidationMode;
import amf.core.client.platform.AMFGraphClient;
import amf.core.client.platform.AMFGraphConfiguration;
import amf.core.client.platform.AMFResult;
import amf.core.client.platform.execution.ExecutionEnvironment;
import amf.core.client.platform.model.document.BaseUnit;
import amf.core.client.platform.model.document.Document;
import amf.core.client.platform.validation.AMFValidationReport;
import amf.shapes.client.platform.model.domain.AnyShape;
import org.junit.Test;

import java.util.concurrent.ExecutionException;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ThreadFactory;

public class SchedulerTest {
    @Test
    public void schedulerTest() throws InterruptedException, ExecutionException {
        /* Instantiating a ScheduledExecutorService with some example arguments. */
        ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(30, new ThreadFactory() {
            int i = 0;

            @Override
            public Thread newThread(Runnable r) {
                i = i + 1;
                final Thread thread = new Thread(r);
                thread.setName("AMF-" + i);
                return thread;
            }
        });

        /* Instantiating a ExecutionEnvironment passing as argument the scheduler previously created. */
        ExecutionEnvironment executionEnvironment = new ExecutionEnvironment(scheduler);

        final AMFGraphConfiguration config = RAMLConfiguration.RAML10()
                .withExecutionEnvironment(executionEnvironment);
        final AMFGraphClient client = config.createClient();

        /* call async interfaces */

        final AMFResult parseResult = client.parse("file://resources/examples/simple-api.raml").get();
        AMFValidationReport report = client.validate(parseResult.baseUnit()).get();
        AMFValidationReport payloadReport = config.payloadValidatorFactory()
                .createFor(obtainShapeFromUnit(parseResult.baseUnit()), "application/json", ValidationMode.StrictValidationMode())
                .validate("{\"name\": \"firstname and lastname\"}").get();
//        config.forInstance

        /* Shutting down the scheduler which kills the AMF threads created in the thread pool provided by that scheduler. */
        scheduler.shutdownNow();
    }

    private AnyShape obtainShapeFromUnit(BaseUnit b) {
        final WebApi webApi = (WebApi) ((Document) b).encodes();
        final EndPoint usersEndpoint = webApi.endPoints().get(0);
        final Operation postMethod = usersEndpoint.operations().get(0);
        final Request request = postMethod.requests().get(0);
        final Payload userPayload = request.payloads().get(0);
        return (AnyShape) userPayload.schema();
    }
}

File: AMF5/src/test/scala/scalaPlatform/SchedulerTest.scala
package scalaPlatform

import amf.apicontract.client.scala.RAMLConfiguration
import amf.apicontract.client.scala.model.domain.api.WebApi
import amf.core.client.common.remote.Content
import amf.core.client.common.validation.ValidationMode
import amf.core.client.scala.execution.ExecutionEnvironment
import amf.core.client.scala.model.document.{BaseUnit, Document}
import amf.core.client.scala.model.domain.Shape
import amf.core.client.scala.resource.{LoaderWithExecutionContext, ResourceLoader}
import org.scalatest.flatspec.AsyncFlatSpec
import org.scalatest.matchers.should

import java.util.concurrent.{Executors, ThreadFactory}
import scala.concurrent.{ExecutionContext, Future}
import scala.io.Source

class SchedulerTest extends AsyncFlatSpec with should.Matchers {

  "AMF configuration" should "allow defining custom execution context" in {
    /* Instantiating a ScheduledExecutorService with some example arguments. */
    val scheduler = Executors.newScheduledThreadPool(30, new ThreadFactory {
      var i: Int = 0
      override def newThread(r: Runnable): Thread = {
        i = i + 1
        val thread = new Thread(r);
        thread.setName("AMF-" + i);
        thread
      }
    });

    /* Instantiating a ExecutionEnvironment passing as argument the scheduler previously created. */
    val executionEnvironment = new ExecutionEnvironment(ExecutionContext.fromExecutorService(scheduler));

    val config = RAMLConfiguration.RAML10()
      .withResourceLoaders(List(CustomResourceLoader(scala.concurrent.ExecutionContext.Implicits.global)))
      .withExecutionEnvironment(executionEnvironment); // execution context of loader is adjusted
    val client = config.createClient();

    /* call async interfaces */
    for {
      parseResult      <- client.parse("file://resources/examples/simple-api.raml")
      validationResult <- client.validate(parseResult.bu)
      payloadReport <- config
        .payloadValidatorFactory()
        .createFor(obtainShapeFromUnit(parseResult.bu), "application/json", ValidationMode.StrictValidationMode)
        .validate("{\"name\": \"firstname and lastname\"}")
    } yield {
      /* Shutting down the scheduler which kills the AMF threads created in the thread pool provided by that scheduler. */
      scheduler.shutdownNow();
      succeed
    }
  }

  def obtainShapeFromUnit(b: BaseUnit): Shape = {
    val webApi = b
      .asInstanceOf[Document]
      .encodes
      .asInstanceOf[WebApi]
    val usersEndpoint = webApi.endPoints.head
    val postMethod    = usersEndpoint.operations.head
    val request       = postMethod.requests.head
    val userPayload   = request.payloads.head
    userPayload.schema
  }

  /*
     resource loader that uses execution context. Extending LoaderWithExecutionContext allows the configuration
     to adapt the loader to the new execution context if modified.
  */
  private case class CustomResourceLoader(ec: ExecutionContext) extends ResourceLoader with LoaderWithExecutionContext {

    override def withExecutionContext(ec: ExecutionContext): ResourceLoader = CustomResourceLoader(ec)

    /** Fetch specified resource and return associated content. Resource should have been previously accepted. */
    override def fetch(resource: String): Future[Content] = {
      val content = Source.fromFile(resource.stripPrefix("file://")).mkString
      Future(new Content(content, resource))(ec)
    }

    /** Accepts specified resource. */
    override def accepts(resource: String): Boolean = true
  }
}
